backend/pcloud/chunkwriter.go

```diff
@@ -0,0 +1,337 @@
+	}
+	openResult, err := fileOpenNew(ctx, client, srcFs, directoryID, leaf)
+	if err != nil {
+		return info, writer, fmt.Errorf("open file: %w", err)
+	}
+
+	info = fs.ChunkWriterInfo{
+		ChunkSize:         int64(srcFs.opt.ChunkSize),
+		Concurrency:       srcFs.opt.UploadConcurrency,
+		LeavePartsOnError: true,
+	}
+	writer = &chunkWriter{
+		client:    client,
+		fs:        srcFs,
+		chunkSize: info.ChunkSize,
+		src:       src,
+		remote:    remote,
+		fd:        openResult.FileDescriptor,
+		fileID:    openResult.Fileid,
+	}
+
+	return info, writer, nil
+}
+
+// Abort implements fs.ChunkWriter.
+func (c *chunkWriter) Abort(ctx context.Context) error {
+	o, err := c.fs.NewObject(ctx, c.remote)
+	if err != nil {
+		return fmt.Errorf("open aborted: %w", err)
+	}
+	if err := o.Remove(ctx); err != nil {
+		return fmt.Errorf("delete aborted: %w", err)
+	}
+	return nil
+}
+
+// Close implements fs.ChunkWriter.
+func (c *chunkWriter) Close(ctx context.Context) error {
+	// In case the file existed previosly and was larger than the newly written file, we need to truncate it to the right size.
+	byteCount := c.GetBytesCount()
+	fs.Debugf(c.src, "truncate target to %d bytes", byteCount)
+	if _, err := c.fileTruncate(ctx, byteCount); err != nil {
+		return fmt.Errorf("truncate file: %w", err)
+	}
+
+	// close fd
+	if _, err := c.fileClose(ctx); err != nil {
+		return fmt.Errorf("close fd: %w", err)
+	}
+
+	// There seems to be no option in the fileops API to set the modtime.
+	// Therefore, we apply the modtime via the regular file API.
+	o, err := c.fs.NewObject(ctx, c.remote)
+	if err != nil {
+		return fmt.Errorf("open newly copied object: %w", err)
+	}
+	if err := o.SetModTime(ctx, c.src.ModTime(ctx)); err != nil {
+		return fmt.Errorf("set modtime: %w", err)
+	}
+
+	return nil
+}
+
+// WriteChunk implements fs.ChunkWriter.
+func (c *chunkWriter) WriteChunk(ctx context.Context, chunkNumber int, reader io.ReadSeeker) (bytesWritten int64, err error) {
+	return c.writeChunk(ctx, chunkNumber, reader)
+}
+
+func (c *chunkWriter) writeChunk(ctx context.Context, chunkNumber int, reader io.Reader) (bytesWritten int64, err error) {
+	chunk := int64(chunkNumber)
+
+	// read chunk into buffer
+	buffer := make([]byte, c.chunkSize)
+	offset := chunk * c.chunkSize
+	eg, egCtx := errgroup.WithContext(ctx)
+
+	// read input buffer and calculate hash
+	inSHA1 := ""
+	eg.Go(func() error {
+		byteCount, err := io.ReadFull(reader, buffer)
+		buffer = buffer[:byteCount]
+		if err != nil && !errors.Is(err, io.ErrUnexpectedEOF) && !errors.Is(err, io.EOF) {
+			return err
+		}
+		inSHA1Bytes := sha1.Sum(buffer)
+		inSHA1 = hex.EncodeToString(inSHA1Bytes[:])
+		return nil
+	})
+
+	// get target hash
+	outSHA1 := ""
+	eg.Go(func() error {
+		targetChecksum, err := c.fileChecksum(egCtx, offset, c.chunkSize)
+		if err != nil {
+			return err
+		}
+		outSHA1 = targetChecksum.SHA1
+		return nil
+	})
+
+	if errors.Is(err, io.EOF) {
```

Comment:
As far as I can see, err hasnt been set here?

Reply:
right, this is why I don't like pre-defined return variables. I'll fix it.